nautobot/core/management/commands/pre_migrate.py

```diff
@@ -75,6 +86,115 @@ def check_configcontext_uniqueness():
         )
 
 
+def check_permissions_constraints(command):
+    """
+    Check for permission constraints that are referencing an object field
+    that will be migrated to a different model/field/value in 2.0 migrations.
+
+    Objects that will be deleted in 2.0 are:
+    - dcim.DeviceRole
+    - dcim.RackRole
+    - ipam.Aggregate
+    - ipam.Role
+
+    Objects that will possibly have fields migrated to a different value in 2.0 are:
+    - ipam.IPAddress
+    - ipam.Prefix
+    - ipam.Service
+    - extras.CustomField
+    - extras.GitRepository
+    - extras.Job
+    - extras.ObjectPermission
+    - extras.ScheduledJob
+    - extras.TaggedItem
+    """
+
+    deleted_models = [
+        Aggregate,
+        DeviceRole,
+        RackRole,
+        Role,
+    ]
+
+    pk_change_models = [
+        IPAddress,
+        Prefix,
+        ObjectPermission,
+        TaggedItem,
+    ]
+
+    field_change_models = {
+        GitRepository: ["slug"],
+        Job: ["commit_default", "name", "module_name", "source", "git_repository"],
+        CustomField: ["label"],
+        ScheduledJob: ["kwargs", "user", "queue", "job_class", "name"],
+        Service: ["name"],
+        ObjectPermission: ["name"],
+    }
+
+    command.stdout.write(command.style.WARNING(">>> Running permission constraint checks..."))
+
+    warnings = []
+
+    for perm in ObjectPermission.objects.all():
+        if not perm.object_types or not perm.constraints:
+            continue
+        constraints = perm.list_constraints()
+        for ct in perm.object_types.all():
+            model = ct.model_class()
+            qs = model.objects.filter(*[models.Q(**c) for c in constraints])
+            if not qs._query.where.children:
+                continue
+
+            for lookup in qs._query.where.children:
+                related_model = lookup.lhs.target.related_model or lookup.lhs.target.model
+
+                if related_model in deleted_models:
+                    cls = f"{related_model.__module__}.{related_model.__name__}"
+                    warnings.append(
+                        f"ObjectPermission '{perm}' (id: {perm.id}) has a constraint that references "
+                        f"a model ({cls}) that will be deleted by the Nautobot 2.0 migration."
```

Comment:
perhaps `deleted or replaced` or something like that? E.g. a constraint on `device_role__name` might just need to become a constraint on `role__name` in 2.0?

Actually, will this catch `device_role__name` or only just `device_role`?

Reply:
This only works on the last model referenced in the qs so `device_role__name` works but it wouldn't catch `device_role__devices__name` because the related model in that case is `Device`. I'm not sure if we can get that granular here without blowing up the scope.